### PR TITLE
[SDF-90] SDF tree component

### PR DESCRIPTION
**Pull request by migoox**
I tried to match the figma desing layout as much as possible, but the full style is a topic of SDF-100.

simplescreenrecorder-2024-12-09_19 46 38

Apart from the component, I refactored the primitive nodes a little:
1. Added unique id for Primitive nodes to support one primitive shader buffer.
2. Reduced boilerplate when defining the primitive.

Also added exhaustive enum switch cases gcc flag, so that the program will not compile on gcc and clang if not all enum options are handled (you may use `default` to bypass it).

Waiting for your feedback 😊

### CODE REVIEW

File: resin/resin/imgui/sdf_tree.cpp [hunk 1]
```diff
@@ -2,140 +2,322 @@
...
 
+  bool is_node_selected = is_parent_selected_ || selected_ == node.node_id();
+
   ImGui::PushID(static_cast<int>(node.node_id().raw()));
+
+  if (is_node_dragged) {
+    ImGui::BeginDisabled();
+  }
+
+  ImGui::PushStyleVar(ImGuiStyleVar_FramePadding, ImVec2(0.f, 4.f));
```
Comment (kubazuch) [inline resin/resin/imgui/sdf_tree.cpp:97]:
```suggestion
  ImGui::PushStyleVar(ImGuiStyleVar_FramePadding, ImVec2(0.F, 4.F));
```

**Review by kubazuch — CHANGES_REQUESTED** [review]

File: libresin/libresin/core/sdf_tree/primitive_base_node.hpp [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+class BasePrimitiveNode;
+using PrimitiveNodeId = Id<BasePrimitiveNode>;
+
+class BasePrimitiveNode : public SDFTreeNode {
+ public:
+  constexpr static auto available_primitive_names() { return kSDFTreePrimitiveNames.names_; }
+
+  virtual SDFTreePrimitiveType primitive_type() const = 0;
+  constexpr std::string_view primitive_name() const { return kSDFTreePrimitiveNames.get_value(primitive_type()); }
```
[the remark this answers is not in the document]
Comment (kubazuch) [reply libresin/libresin/core/sdf_tree/primitive_base_node.hpp:20]: The issue is probably related to `constexpr primitive_name()` calling `virtual primitive_type()`

[the remark this answers is not in the document]
Comment (migoox) [reply libresin/libresin/core/sdf_tree/primitive_base_node.hpp]: Making the `primitive_name()` a virtual function and moving the definition to children fixed the issue

File: resin/resin/imgui/sdf_tree.cpp [hunk 3]
```diff
@@ -2,140 +2,322 @@
...
 }
 
-void SDFTreeOperationVisitor::visit_primitive(::resin::PrimitiveNode& node) {
-  if (op == SDFTreeOperationVisitor::Operation::PushGroup) {
-    node.parent()
-        .push_back_child<::resin::GroupNode>(::resin::SDFBinaryOperation::Union)
-        .push_back_child<::resin::SphereNode>(::resin::SDFBinaryOperation::Union);
-  } else if (op == SDFTreeOperationVisitor::Operation::PushPrimitive) {
-    node.parent().push_back_child<::resin::SphereNode>(::resin::SDFBinaryOperation::Union);
+void SDFTreeComponentVisitor::render_tree(::resin::SDFTree& tree) {
```
Comment (kubazuch) [inline resin/resin/imgui/sdf_tree.cpp:199]: Some comments on behavior:
- [ ] minimal width and height
- [ ] some way to remove selections, i.e. by clicking the background of node list
- [ ] there seems to be some problem regarding padding (fig1)

Fig1:
image

Comment (migoox) [reply resin/resin/imgui/sdf_tree.cpp:199]: > minimal width and height
Sadly I couldn't find any way to define the constraints for components and enforce them on the windows. For that reason we should constraint the windows instead:(

**Review by kubazuch — APPROVED** [review]

**Review by adamieck — APPROVED** [review]

**Pull request merged**